CORE-788: Quasar instrumentation should not reference classes not in the agent.

chrisr3 opened the pull request:
Refactor classes used in Quasar instrumentation into a separate package inside the agent jar itself. We can then expose this _single_ package across an entire OSGi framework via "boot-delegation".

Note that OSGi must actually delegate to the application classloader and not the bootstrap classloader for this to work. This requires us also to configure:
```
org.osgi.framework.bundle.parent=app
```

PaulHatcherR3 commented:
Noted that you've added fibers.suspend and added in the interface stackOps to dance across packages I assume ?

As you've now moved the stack to an interface does this have much impact on runtime performance ?

chrisr3 commented:
> Noted that you've added fibers.suspend and added in the interface stackOps to dance across packages I assume ?

This new package contains the public instrumentation code that the Quasar agent will inject into each class. OSGi works at the "package" level, and so we can configure the OSGi framework to "boot-delegate" _only_ this package. I have fought valiantly to preserve the existing `public` / `protected` / `private` relationships between the fibers' existing classes. It's still the exact same `Stack` object being passed around as before, except the instrumentation code is now accessing it via an interface. This means that the instrumentor no longer injects class references that the OSGi "bundle wiring" doesn't know how to resolve, which was previously causing `ClassNotFoundException` at runtime.

> As you've now moved the stack to an interface does this have much impact on runtime performance ?

I obviously haven't "performance tested" this. But I cannot imagine **any** performance impact being so terrible that we'd prefer _not_ to allow Quasar to use OSGi boot delegation. When compared with the performance where loading instrumented classes fails with `ClassNotFoundException`, I would estimate that we actually have an _infinite improvement_ :wink:.

> Also, in the files you've added as this is a public repository probably need to copy the license comment header.

I can add some, although they're obviously new files that I've written myself.

rick-r3 commented on quasar-core/src/main/java/co/paralleluniverse/fibers/suspend/Stacks.java:
```diff
@@ -0,0 +1,26 @@
+
+@SuppressWarnings("unused")
+public final class Stacks {
+    private static volatile Supplier<? extends StackOps> GET_STACK;
```
I think the memory model says that this doesn't have to be `volatile` if the threads that call it are created after it is set.  So we could remove the `volatile` if we had an "initialise quasar" call that we did somewhere during node start (and it probably works accidentally already without it).   I wonder to what extent we should bother or is this code riddled with memory barriers anyway?  I guess a suspension point is expensive anyway, so perhaps no need to bother.

chrisr3 replied:
I was preferring the code to be correct in its own right rather than relying on having "magic" code in a client. However, on the subject of performance, I have found [this article](https://dzone.com/articles/invoke-interface-optimisations) that suggests `invokeinterface` can be slower than `invokevirtual` - although only the method look-up should still be very simple.

However, if you promise _not_ to tell @mnesbit then I can recreate `StackOps` as an `abstract` base class so that the `invokeinterface` instructions can become `invokevirtual` again.

rick-r3 replied:
I was going to say, is `StackOps` as abstract possible?  Could also then undo java version 8 stuff.  It can have the static `StackOps.getStack()` too, then ? And use `protected` for the static so that `Stack` can initialise?

chrisr3 replied:
Most of that is easy, but I don't believe having
```java
protected static volatile Supplier<? extends StackOps> GET_STACK;
```
is a good idea. If `Stack` can modify it without using reflection then so can
```java
public class EvilStack extends StackOps {}
```
Remember that `StackOps` _**must**_ be`public`, non-`final`. _**and**_ globally available throughout the OSGi framework! :scream:

rick-r3 replied:
I think there are mechanisms to eliminate the `volatile` since `MethodHandles` are immutable, but we go down a bit of a rabbit hole of edge cases that probably isn't worth it unless we really run into issues with the `volatile` in the future when looking at performance.

rick-r3 commented on quasar-core/src/main/java/co/paralleluniverse/fibers/suspend/Stacks.java:
```diff
@@ -0,0 +1,26 @@
+
+    public static StackOps getStack() {
+        final Supplier<? extends StackOps> getter = GET_STACK;
+        return getter != null ? getter.get() : null;
```
I wonder to what extent it makes sense to  `throw RuntimeException("Quasar not correctly initialised")` rather than returning null ?  Does it actually surface I wonder or lost in the wilderness of a weird Quasar NPEs etc?

chrisr3 replied:
The underlying `Stack.getStack()` function can return `null`:
```java
public static Stack getStack() {
    final Fiber<?> currentFiber = Fiber.currentFiber();
    return currentFiber != null ? currentFiber.stack : null;
}
```
and so I don't believe that `Stacks.getStack()` also being able to return `null` should be a problem.

rick-r3 replied:
Well, we have seen really annoying NPEs when quasar gets a bit confused and wondered if it was an opportunity to tidy that up a bit, at least where the initialisation hasn't worked.

chrisr3 replied:
The instrumentation code does check whether `getStack()` returns `null`. Did you trace your NPEs back to `Stack` operations?

There _will_ now be a delay before the OSGi framework can load the `quasar-core-osgi` bundle, during which time the agent will still be instrumenting classes. I think that handling this period gracefully would be a Good Thing :tm:.

rick-r3 replied:
Yes, I'd not thought about the fact that instrumented code is executed (most likely) on Threads before any Fibers exist.  Good point and my suggestion would probably break Corda.  Will mark as resolved!

mnesbit reviewed (approved):
OK
